internal/server/data/provider.go

```diff
@@ -20,32 +40,147 @@ func validateProvider(p *models.Provider) error {
+	// created by this user. Can be used with NotIDs.
+	CreatedBy uid.ID
+	// NotIDs instructs DeleteProviders to exclude any providers with these IDs to
+	// be excluded. In other words, these IDs will not be deleted, even if they
+	// match CreatedBy.
+	// Can only be used with CreatedBy.
+	NotIDs []uid.ID
+
+	Pagination *Pagination
 }
 
-func ListProviders(db GormTxn, p *Pagination, selectors ...SelectorFunc) ([]models.Provider, error) {
-	return list[models.Provider](db, p, selectors...)
+func ListProviders(tx ReadTxn, opts ListProvidersOptions) ([]models.Provider, error) {
+	table := providersTable{}
+	query := querybuilder.New("SELECT")
+	query.B(columnsForSelect(table))
+	if opts.Pagination != nil {
+		query.B(", count(*) OVER()")
+	}
+	query.B("FROM providers")
+	query.B("WHERE deleted_at is null")
+	query.B("AND organization_id = ?", tx.OrganizationID())
+
+	if opts.ByName != "" {
+		query.B("AND name = ?", opts.ByName)
+	}
+	if opts.ExcludeInfraProvider {
+		query.B("AND kind <> ?", models.ProviderKindInfra)
+	}
+	if len(opts.ByIDs) > 0 {
+		query.B("AND id IN (?)", opts.ByIDs)
+	}
+	if opts.CreatedBy != 0 {
+		query.B("AND created_by = ?", opts.CreatedBy)
+		if len(opts.NotIDs) > 0 {
+			query.B("AND id NOT IN (?)", opts.NotIDs)
+		}
+	}
+
+	query.B("ORDER BY name ASC")
+	if opts.Pagination != nil {
+		opts.Pagination.PaginateQuery(query)
+	}
+
+	rows, err := tx.Query(query.String(), query.Args...)
+	if err != nil {
+		return nil, err
+	}
+	return scanRows(rows, func(grant *models.Provider) []any {
+		fields := (*providersTable)(grant).ScanFields()
+		if opts.Pagination != nil {
+			fields = append(fields, &opts.Pagination.TotalCount)
+		}
+		return fields
+	})
 }
 
-func SaveProvider(db GormTxn, provider *models.Provider) error {
+func UpdateProvider(tx WriteTxn, provider *models.Provider) error {
 	if err := validateProvider(provider); err != nil {
 		return err
 	}
-	return save(db, provider)
+	return update(tx, (*providersTable)(provider))
 }
 
-func DeleteProviders(db GormTxn, selectors ...SelectorFunc) error {
-	toDelete, err := ListProviders(db, nil, selectors...)
-	if err != nil {
-		return fmt.Errorf("listing providers: %w", err)
+type DeleteProvidersOptions struct {
+	// ByID instructs DeleteProviders to delete the provider matching this ID.
+	// When non-zero all other fields on this struct will be ignored
+	ByID uid.ID
+
+	// CreatedBy instructs DeleteProviders to delete all the providers that were
+	// created by this user. Can be used with NotIDs.
+	CreatedBy uid.ID
+	// NotIDs instructs DeleteProviders to exclude any providers with these IDs to
+	// be excluded. In other words, these IDs will not be deleted, even if they
+	// match CreatedBy.
+	// Can only be used with CreatedBy.
+	NotIDs []uid.ID
+}
+
+func DeleteProviders(db GormTxn, opts DeleteProvidersOptions) error {
+	var toDelete []models.Provider
+	if opts.ByID != 0 {
+		if provider, _ := GetProvider(db, GetProviderOptions{ByID: opts.ByID}); provider != nil {
```

Comment:
Why not just add this ID directly to `toDelete`?

Reply:
Great question! I had it that way originally, but changed it in e8cc797acacf7e84a2b85d2a83c2d8770f784f44.

The problem with just using the ID was that if someone guessed an ID it would allow them to delete some provider_user entries from a different org.

I think another way to deal with this would be to first delete the provider, and use `RETURNING id`. That would give us all the provider IDs instead of having to call `ListProviders`. I'll explore that option in a follow up PR.